File: Java/gray-code.java
/**
 * Problem Link: https://leetcode.com/problems/gray-code/
 * 
 * Iterate every code, generate next code based on current(latest) code.
 */
public class Solution {
    private HashSet<Integer> set;
    private List<Integer> list;
    
    public List<Integer> grayCode(int n) {
        set = new HashSet<Integer>();
        list = new ArrayList<Integer>();
        set.add(0);
        list.add(0);
        next(0, n);
        return list;
    }
    
    private void next(int last, int n) {
        for(int i=0; i<n; i++) {
            int bit = 1 & (last >> i);
            int ret = last;
            if(bit == 0) {
                ret += (1 << i);
            }
            else {
                ret -= (1 << i);
            }
            //System.out.println(ret);
            Integer r = Integer.valueOf(ret);
            if(!set.contains(r)) {
                list.add(r);
                set.add(r);
                next(ret, n);
            }
        }
    }
}

File: Swift/gray-code.swift
/**
 * https://leetcode.com/problems/gray-code/
 * 
 * 
 */ 
// Date: Thu Jul  1 10:20:34 PDT 2021
class Solution {
    func grayCode(_ n: Int) -> [Int] {
        var result = [0, 1]
        for offset in stride(from: 1, to: n, by: 1) {
            let addon = 1 << offset
            var list = [Int]()
            for index in stride(from: result.count - 1, through: 0, by: -1) {
                list.append(result[index] + addon)
            }
            result += list
        }
        return result
    }
}